doc: add sign-txn page (#15103)
MystenLabs/sui: docs/content/guides/developer/sui-101/sign-and-send-txn.mdx

* Construct the transaction data by creating a `TransactionBlock` where multiple transactions are chained. See [Building Programmable Transaction Blocks](./building-ptb.mdx) for more information. 
* The SDK's built-in gas estimation and coin selection picks the gas coin.
* Sign the transaction to generate a [signature](/concepts/cryptography/transaction-auth/signatures#user-signature).
* Submit the `TransactionBlock` and its signature for on-chain execution.

:::info

If you want to use a specific gas coin, first find the gas coin object ID to be used to pay for gas, and explicitly use that in the PTB. If there is no gas coin object, use the [splitCoin](./building-ptb#available-transactions) transaction to create a gas coin object. The split coin transaction should be the first transaction call in the PTB.

:::

## Examples

The following examples demonstrate how to sign and execute transactions using Rust, TypeScript, or the Sui CLI.

<Tabs groupId="code-language">

<TabItem value="ts-sdk" label="TypeScript">

There are various ways to instantiate a key pair and to derive its public key and Sui address using the Sui TypeScript SDK.

```tsx

>     import { Ed25519Keypair } from '@mysten/sui.js/keypairs/ed25519';
>     import { Secp256k1Keypair } from '@mysten/sui.js/keypairs/secp256k1';
>     import { Secp256r1Keypair } from '@mysten/sui.js/keypairs/secp256r1';
>     import { fromHEX } from '@mysten/bcs';
>     import { SuiClient, getFullnodeUrl } from '@mysten/sui.js/client';
>     import { type Keypair } from '@mysten/sui.js/cryptography';
>     import { TransactionBlock } from '@mysten/sui.js/transactions';

Reviewer: ```suggestion
    import { SuiClient, getFullnodeUrl } from '@mysten/sui.js/client';
    import { TransactionBlock } from '@mysten/sui.js/transactions';
    import { type Keypair } from '@mysten/sui.js/cryptography';
    import { Ed25519Keypair } from '@mysten/sui.js/keypairs/ed25519';
    import { Secp256k1Keypair } from '@mysten/sui.js/keypairs/secp256k1';
    import { Secp256r1Keypair } from '@mysten/sui.js/keypairs/secp256r1';
    import { fromHEX } from '@mysten/bcs';
```

    const kp_import_0 = Ed25519Keypair.fromSecretKey(fromHex("0xd463e11c7915945e86ac2b72d88b8190cfad8ff7b48e7eb892c275a5cf0a3e82"));
    const kp_import_1 = Secp256k1Keypair.fromSecretKey(fromHex("0xd463e11c7915945e86ac2b72d88b8190cfad8ff7b48e7eb892c275a5cf0a3e82"));
    const kp_import_2 = Secp256r1Keypair.fromSecretKey(fromHex("0xd463e11c7915945e86ac2b72d88b8190cfad8ff7b48e7eb892c275a5cf0a3e82"));

	const kp_derive_0 = Ed25519Keypair.deriveKeypair("$MNEMONICS");
	const kp_derive_1 = Secp256k1Keypair.deriveKeypair("$MNEMONICS");
	const kp_derive_2 = Secp256r1Keypair.deriveKeypair("$MNEMONICS");

	const kp_derive_with_path_0 = Ed25519Keypair.deriveKeypair("$MNEMONICS", "m/44'/784'/1'/0'/0'");
	const kp_derive_with_path_1 = Secp256k1Keypair.deriveKeypair("$MNEMONICS", "m/54'/784'/1'/0/0");
	const kp_derive_with_path_2 = Secp256r1Keypair.deriveKeypair("$MNEMONICS", "m/74'/784'/1'/0/0");

    // replace `kp_rand_0` with the variable names above.
    const pk = kp_rand_0.getPublicKey();
    const sender = pk.toSuiAddress();

    // create an example transaction block. 
    const txb = new TransactionBlock();
	txb.setSender(sender);
	txb.setGasPrice(5);

> 	txb.setGasBudget(100);

Reviewer: is this in MIST?

Author: im trying to focus on the sign/execute tx part without going into the details here that can be distracting. this is just an example tx as mentioned, we discussed at the beginning that this for programmable txn building details refer to the other page.

	const bytes = await txb.build();
    const serializedSignature = (await keypair.signTransactionBlock(bytes)).signature;

>     // verify the signature locally. 

Reviewer: What do we mean by locally? Can we do it somewhere else, not locally?

Author: verifying it locally is the right thing to do

Author: im confused about the suggestion. this should be interpreted as: if you would like to know how to locally verify signature (frankly a lot of devs ask about this), here is the sdk on how to do it. if you dont want to do it or dont care what this means, or dont want to do it, its fine to skip.

        )
        .unwrap(),
    );
    let _skp_import_no_flag_2 = SuiKeyPair::Ed25519(
        Ed25519KeyPair::from_bytes(
            &Base64::decode("1GPhHHkVlF6GrCty2IuBkM+tj/e0jn64ksJ1pc8KPoI=").unwrap(),
        )
        .unwrap(),
    );

    // import a keypair from a base64 encoded 33-byte `flag || private key`. The signature scheme is determined by the flag.
    let _skp_import_with_flag_0 =
        SuiKeyPair::decode_base64("ANRj4Rx5FZRehqwrctiLgZDPrY/3tI5+uJLCdaXPCj6C").unwrap();
    let _skp_import_with_flag_1 =
        SuiKeyPair::decode_base64("AdRj4Rx5FZRehqwrctiLgZDPrY/3tI5+uJLCdaXPCj6C").unwrap();
    let _skp_import_with_flag_2 =
        SuiKeyPair::decode_base64("AtRj4Rx5FZRehqwrctiLgZDPrY/3tI5+uJLCdaXPCj6C").unwrap();

    // replace `skp_determ_0` with the variable names above
    let pk = skp_determ_0.public();
    let sender = SuiAddress::from(&pk);
```

Next, sign transaction data constructed using an example programmable transaction block with default gas coin, gas budget, and gas price. See [Building Programmable Transaction Blocks](./building-ptb.mdx) for more information. 

> ```

Reviewer: ```suggestion
```rust
```

    // construct an example programmable transaction. 
    let pt = {
        let mut builder = ProgrammableTransactionBuilder::new();

>         builder.pay_sui(vec![sender], vec![1]).unwrap();

Reviewer: let's explain what `vec[1]` means. Also, no `unwrap()` please.

        builder.finish()
    };

    let gas_budget = 5_000_000;
    let gas_price = sui_client.read_api().get_reference_gas_price().await?;

    // create the transaction data that will be sent to the network.
    let tx_data = TransactionData::new_programmable(
        sender,
        vec![gas_coin.object_ref()],
        pt,
        gas_budget,
        gas_price,
    );
```

Commit a signature to the Blake2b hash digest of the intent message (`intent || bcs bytes of tx_data`). 

```Rust
    // derive the digest that the keypair should sign on. 
    let intent_msg = IntentMessage::new(Intent::sui_transaction(), tx_data);
    let raw_tx = bcs::to_bytes(&intent_msg).map_err(|_| TransactionError::ParsingError)?;
    let mut hasher = sui_types::crypto::DefaultHash::default();
    hasher.update(raw_tx.clone());
    let digest = hasher.finalize().digest;

    // use SuiKeyPair to sign the digest. 
    let sui_sig = skp.sign(&digest);

>     // verify the signature against the transaction locally. if this fails the transaction probably will fail submitting. 

Reviewer: ```suggestion
    // verify the signature against the transaction locally. If this fails, the transaction will probably fail submitting. 
```
What do we mean by `fail submitting` - is it `fail to be submitted to be executed?` or it will simply fail and an error will be return. Let's be precise here, it is important to accurately describe it.

Author: im not sure i understand the difference. you can submit whatever you want, but if an invalid signature is submitted, it will return a Signature Error, which is why i provided verify signature locally function bc ppl do like figure this out locally.

Reviewer: Well, we are a bit vague: if the signature is invalid, then it will return an error and thus there's no `probably will fail submitting`, right?

    let res = sui_sig.verify_secure(&intent_msg, sender, sui_types::crypto::SignatureScheme::ED25519);
    assert!(res.is_ok());
```

Finally, submit the transaction with the signature. 

```Rust
    let transaction_response = sui_client
        .quorum_driver_api()
        .execute_transaction_block(
            sui_types::transaction::Transaction::from_generic_sig_data(
                intent_msg.value,
                Intent::sui_transaction(),
                vec![GenericSignature::Signature(sui_sig)],
            ),
            SuiTransactionBlockResponseOptions::default(),
            None,
        )
        .await?;

> 

Reviewer: nit: remove empty line

```

</TabItem>

<TabItem value="cli" label="Sui CLI">

> The Sui CLI initiates a local file in `~/.sui/keystore` on your machine with a list of private keys (encoded as Base64 encoded `flag || 32-byte-private-key`) that you can use to sign transactions by specifying its address.

Reviewer: ```suggestion
When using the Sui CLI for the first time, it creates a local file in `~/.sui/keystore` on your machine with a list of private keys (encoded as Base64 encoded `flag || 32-byte-private-key`). You can use any key to sign transactions by specifying its address. Use `sui keytool list` to see a list of addresses.
```

There are three ways to initialize a key:

```shell
# generate randomly. 
sui client new-address ed25519
sui client new-address secp256k1
sui client new-address secp256r1

# import the 32-byte private key to keystore.
sui keytool import "0xd463e11c7915945e86ac2b72d88b8190cfad8ff7b48e7eb892c275a5cf0a3e82" ed25519
sui keytool import "0xd463e11c7915945e86ac2b72d88b8190cfad8ff7b48e7eb892c275a5cf0a3e82" secp256k1
sui keytool import "0xd463e11c7915945e86ac2b72d88b8190cfad8ff7b48e7eb892c275a5cf0a3e82" secp256r1

# import the mneomonics with derivation path to keystore. 
sui keytool import "$MNEMONICS" ed25519
sui keytool import "$MNEMONICS" secp256k1
sui keytool import "$MNEMONICS" secp256r1
```

> Create a transfer transaction in CLI. Set the $SUI_ADDRESS to the one corresponding to the key pair used to sign. Then sign with the private key corresponding to the sender address. 

Reviewer: We should also mention about $GAS_COIN_ID and $GAS_BUDGET variables.

```shell
sui client gas
sui client transfer-sui --to $SUI_ADDRESS --sui-coin-object-id $GAS_COIN_ID --gas-budget $GAS_BUDGET --serialize-unsigned-transaction
sui keytool sign --address $SUI_ADDRESS --data $TX_BYTES
sui client execute-signed-tx --tx-bytes $TX_BYTES --signatures $SERIALIZED_SIGNATURE
```

</TabItem>
</Tabs>

### Notes
1. This guide demonstrates how to sign with a single private key. Refer to [Multisig](/concepts/cryptography/transaction-auth/multisig.mdx) when it is preferred to set up more complex signing policies.

> 2. Similarly, native zkLogin does not follow the above steps, see [zkLogin](/concepts/cryptography/zklogin.mdx) to understand how to derive a zkLogin address, produce a zkLogin signature with an ephemeral key pair. 

Reviewer: ```suggestion
2. Similarly, native zkLogin does not follow the above steps, see [the docs](/concepts/cryptography/zklogin.mdx) to understand how to derive a zkLogin address, and produce a zkLogin signature with an ephemeral key pair. 
```

> 3. If you decide to implement your own signing mechanisms instead of using the previous tools, see [Sui Signatures](/concepts/cryptography/transaction-auth/signatures.mdx) on the accepted signature specifications for each scheme. 

Reviewer: ```suggestion
3. If you decide to implement your own signing mechanisms instead of using the previous tools, see the [Signatures](/concepts/cryptography/transaction-auth/signatures.mdx) doc on the accepted signature specifications for each scheme. 
```

> 4. Flag is one byte that differentiates signature schemes. See supported schemes and its flag in [Sui Signatures](/concepts/cryptography/transaction-auth/signatures.mdx).

Reviewer: ```suggestion
4. Flag is one byte that differentiates signature schemes. See supported schemes and its flag in [Signatures](/concepts/cryptography/transaction-auth/signatures.mdx).
```

> 5. The `execute_transaction_block` endpoint takes a list of signatures, it should contain exactly one user signature, unless you are using sponsored transaction that a second signature for the gas object can be provided. See [Sponsored Transactions](concepts/transactions/sponsored-transactions.mdx) for more information. 

Reviewer: ```suggestion
5. The `execute_transaction_block` endpoint takes a list of signatures, so it should contain exactly one user signature, unless you are using sponsored transaction that a second signature for the gas object can be provided. See [Sponsored Transactions](/concepts/transactions/sponsored-transactions.mdx) for more information. 
```

---
title: Signing and Sending Transactions
---

import Tabs from '@theme/Tabs';
import TabItem from '@theme/TabItem';

Transactions in Sui represent calls to specific functionality (like calling a smart contract function) that execute on inputs to define the result of the transaction.

Inputs can either be an object reference (either to an owned object, an immutable object, or a shared object), or an encoded value (for example, a vector of bytes used as an argument to a Move call). After a transaction is constructed, usually through using [programmable transaction blocks](./building-ptb) (PTBs), the user signs the transaction and submits it to be executed on chain.

> The signature is provided with the private key owned by the wallet, and its public key must be consistent with the transaction sender Sui address. 

Reviewer: ```suggestion
The signature is provided with the private key owned by the wallet, and its public key must be consistent with the transaction sender's Sui address. 
```

Sui uses a `SuiKeyPair` to produce the signature, which commits to the Blake2b hash digest of the intent message (`intent || bcs bytes of tx_data`). The signature schemes currently supported are Ed25519 Pure, ECDSA Secp256k1, ECDSA Secp256r1, Multisig, and ZkLogin. 

> You can instantiate Ed25519 Pure, ECDSA Secp256k1, and ECDSA Secp256r1 using `SuiKeyPair` and use it to sign transactions. Note that this guide does not apply to multisig and ZkLogin, please refer to their own pages for instructions. 

Reviewer: ```suggestion
You can instantiate `Ed25519 Pure`, `ECDSA Secp256k1`, and `ECDSA Secp256r1` using `SuiKeyPair` and use it to sign transactions. Note that this guide does not apply to `Multisig` and `zkLogin`, please refer to their own pages ([Multisig](/concepts/cryptography/transaction-auth/multisig) and [zkLogin](https://docs.sui.io/concepts/cryptography/zklogin) respectively) for instructions. 
```

> With a signature and the transaction bytes, a transaction can be executed successfully. 

Reviewer: ```suggestion
With a signature and the transaction bytes, a transaction can now be submitted to be executed.
```

<TabItem value="rust" label="Rust">

There are various ways to instantiate a `SuiKeyPair` and to derive its public key and Sui address using the Sui Rust SDK. You can find the complete runnable in [Sui repository](https://github.com/MystenLabs/sui/tree/main/crates/sui-sdk/sign_tx_guide.rs)/

```Rust

>     // deterministically generate a keypair, testing only, do not use for mainnet.

Reviewer: In the text we use `key pair` and here we have `keypair`. Make it consistent.
